kotlin/android/app/src/main/java/dev/firezone/android/features/session/ui/SessionActivity.kt

```diff
@@ -98,12 +125,36 @@ internal class SessionActivity : AppCompatActivity() {
             }
         }
 
-        viewModel.resourcesLiveData.observe(this) { resources ->
-            resourcesAdapter.submitList(resources)
+        viewModel.resourcesLiveData.observe(this) { value ->
+            resourceList = value
+            refreshList()
+        }
+
+        viewModel.favoriteResources.observe(this) {
+            refreshList()
+        }
+        viewModel.favoriteResources.value = viewModel.repo.getFavoritesSync()
+    }
+
+    private fun refreshList() {
+        if (viewModel.favoriteResources.value!!.isEmpty()) {
+            showOnlyFavorites = false
+            binding.tabLayout.selectTab(binding.tabLayout.getTabAt(RESOURCES_TAB_ALL), true)
+            binding.tabLayout.visibility = View.GONE
+            resourcesAdapter.submitList(resourceList)
+        } else if (showOnlyFavorites) {
+            val list = resourceList.filter { viewModel.favoriteResources.value!!.contains(it.id) }
+            binding.tabLayout.visibility = View.VISIBLE
+            resourcesAdapter.submitList(list)
+        } else {
+            binding.tabLayout.visibility = View.VISIBLE
+            resourcesAdapter.submitList(resourceList)
```

Comment:
UIs are notoriously hard to test. View-models help with that because they can act as a place where you put all your logic in what should get displayed. Consequently, the actual code that changes the UI is so "stupid" that you don't need to test it.

Instead of this conditional logic here, I'd suggest to create functions on `viewModel` that compute you the things that you need:

- A getter for `showOnlyFavourites`
- A getter for the tab visibility
- A getter for the tab index
- A getter for which resources are sent to the `resourcesAdapter`